### DISCUSSION TITLE
Can I use the stress branch for training, and main branch for deploying and evaulation?

### DISCUSSION THREAD

**Discussion by phys-chem** [mir-group/nequip#412] [Q&A]
I need to run NPT simulations but meeting problems in installing torch-scatter. When I use the stress branch, I need torch-scatter for deploying. However, I switch the main branch, it cannot be deployed. It is right? Thank you very much!

I meet this error when using main branch for model trained by stress branch:
RuntimeError: Error(s) in loading state_dict for RescaleOutput:
	Missing key(s) in state_dict: "scale_by", "shift_by", "model._empty", "model.func.radial_basis.basis.bessel_weights", "model.func.chemical_embedding.linear.weight", "model.func.chemical_embedding.linear.bias", "model.func.chemical_embedding.linear.output_mask", "model.func.layer0_convnet.equivariant_nonlin.mul.weight", "model.func.layer0_convnet.equivariant_nonlin.mul.output_mask", "model.func.layer0_convnet.conv.linear_1.weight", "model.func.layer0_convnet.conv.linear_1.bias", "model.func.layer0_convnet.conv.linear_1.output_mask", "model.func.layer0_convnet.conv.fc.layer0.weight", "model.func.layer0_convnet.conv.fc.layer1.weight", "model.func.layer0_convnet.conv.fc.layer2.weight", "model.func.layer0_convnet.conv.tp.weight", "model.func.layer0_convnet.conv.tp.output_mask", "model.func.layer0_convnet.conv.linear_2.weight", "model.func.layer0_convnet.conv.linear_2.bias", "model.func.layer0_convnet.conv.linear_2.output_mask", "model.func.layer0_convnet.conv.sc.weight", "model.func.layer0_convnet.conv.sc.output_mask", "model.func.layer1_convnet.equivariant_nonlin.mul.weight", "model.func.layer1_convnet.equivariant_nonlin.mul.output_mask", "model.func.layer1_convnet.conv.linear_1.weight", "model.func.layer1_convnet.conv.linear_1.bias", "model.func.layer1_convnet.conv.linear_1.output_mask", "model.func.layer1_convnet.conv.fc.layer0.weight", "model.func.layer1_convnet.conv.fc.layer1.weight", "model.func.layer1_convnet.conv.fc.layer2.weight", "model.func.layer1_convnet.conv.tp.weight", "model.func.layer1_convnet.conv.tp.output_mask", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_1_1_1", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_1_1_2", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_1_2_1", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_1_2_2", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_2_1_1", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_2_1_2", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_2_2_1", "model.func.layer1_convnet.conv.tp._compiled_main_left_right._w3j_2_2_2", "model.func.layer1_convnet.conv.linear_2.weight", "model.func.layer1_convnet.conv.linear_2.bias", "model.func.layer1_convnet.conv.linear_2.output_mask", "model.func.layer1_convnet.conv.sc.weight", "model.func.layer1_convnet.conv.sc.output_mask", "model.func.layer2_convnet.equivariant_nonlin.mul.weight", "model.func.layer2_convnet.equivariant_nonlin.mul.output_mask", "model.func.layer2_convnet.conv.linear_1.weight", "model.func.layer2_convnet.conv.linear_1.bias", "model.func.layer2_convnet.conv.linear_1.output_mask", "model.func.layer2_convnet.conv.fc.layer0.weight", "model.func.layer2_convnet.conv.fc.layer1.weight", "model.func.layer2_convnet.conv.fc.layer2.weight", "model.func.layer2_convnet.conv.tp.weight", "model.func.layer2_convnet.conv.tp.output_mask", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_1_1_1", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_1_1_2", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_1_2_1", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_1_2_2", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_2_1_1", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_2_1_2", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_2_2_1", "model.func.layer2_convnet.conv.tp._compiled_main_left_right._w3j_2_2_2", "model.func.layer2_convnet.conv.linear_2.weight", "model.func.layer2_convnet.conv.linear_2.bias", "model.func.layer2_convnet.conv.linear_2.output_mask", "model.func.layer2_convnet.conv.sc.weight", "model.func.layer2_convnet.conv.sc.output_mask", "model.func.layer3_convnet.equivariant_nonlin.mul.weight", "model.func.layer3_convnet.equivariant_nonlin.mul.output_mask", "model.func.layer3_convnet.conv.linear_1.weight", "model.func.layer3_convnet.conv.linear_1.bias", "model.func.layer3_convnet.conv.linear_1.output_mask", "model.func.layer3_convnet.conv.fc.layer0.weight", "model.func.layer3_convnet.conv.fc.layer1.weight", "model.func.layer3_convnet.conv.fc.layer2.weight", "model.func.layer3_convnet.conv.tp.weight", "model.func.layer3_convnet.conv.tp.output_mask", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_1_1_1", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_1_1_2", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_1_2_1", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_1_2_2", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_2_1_1", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_2_1_2", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_2_2_1", "model.func.layer3_convnet.conv.tp._compiled_main_left_right._w3j_2_2_2", "model.func.layer3_convnet.conv.linear_2.weight", "model.func.layer3_convnet.conv.linear_2.bias", "model.func.layer3_convnet.conv.linear_2.output_mask", "model.func.layer3_convnet.conv.sc.weight", "model.func.layer3_convnet.conv.sc.output_mask", "model.func.conv_to_output_hidden.linear.weight", "model.func.conv_to_output_hidden.linear.bias", "model.func.conv_to_output_hidden.linear.output_mask", "model.func.output_hidden_to_scalar.linear.weight", "model.func.output_hidden_to_scalar.linear.bias", "model.func.output_hidden_to_scalar.linear.output_mask", "model.func.per_species_rescale.shifts", "model.func.per_species_rescale.scales".

**Comment by Linux-cpp-lisp** [comment]
Hi @phys-chem ,

I suggest using the `develop` branch, which should not have this dependency on `torch-scatter`.  Thanks for your interest in our code!

**Reply by phys-chem** [reply]
Thank you! Therefore, should I use develop branch for training NequiP, and stress branch for pair-nequip for running NPT?

**Reply by Linux-cpp-lisp** [reply]
If you want to deploy a NequIP (not Allegro) model to `pair_nequip`, you may need to set `default_dtype: float32`  on `nequip` `develop`

**Reply by Linux-cpp-lisp** [reply]
You could also try `nequip` `develop` + `allegro` + `pair_allegro`

**Reply by phys-chem** [reply]
Thank you very much! Do I need to use float64 for running NPT?

**Reply by Linux-cpp-lisp** [reply]
On `nequip` `develop`, I would recommend it. (`default_dtype: float64`, as is the default, I mean)
